pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2024 MongoDB.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package atlasprivateendpoint
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/privateendpoint"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+)
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointService(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+) (ctrl.Result, error) {
+	akoPEService := privateendpoint.NewPrivateEndpoint(akoPrivateEndpoint)
+	atlasPEService, err := r.getPrivateEndpointService(ctx.Context, projectID, akoPrivateEndpoint)
+	if err != nil {
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.ReadyType, workflow.Internal, err)
+	}
+
+	wasDeleted := !akoPrivateEndpoint.GetDeletionTimestamp().IsZero()
+	existInAtlas := atlasPEService != nil
+
+	switch {
+	case !existInAtlas && !wasDeleted:
+		atlasPEService, err = r.privateEndpointService.CreatePrivateEndpointService(ctx.Context, projectID, akoPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToCreate, err)
+		}
+	case !existInAtlas && wasDeleted:
+		return r.unmanage(ctx, akoPrivateEndpoint)
+	case existInAtlas && wasDeleted:
+		if customresource.IsResourcePolicyKeepOrDefault(akoPrivateEndpoint, r.ObjectDeletionProtection) {
+			return r.unmanage(ctx, akoPrivateEndpoint)
+		}
+
+		atlasPEService, err = r.deletePrivateEndpoint(ctx.Context, projectID, atlasPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointFailedToDelete, err)
+		}
+	}
+
+	switch atlasPEService.Status() {
+	case privateendpoint.StatusInitiating:
```

Comment:
I am confused here. Maybe we should document this state machine somewhere else.

The first switch handles synchronising the existence of the private endpoint service between Atlas and the Kubernetes spec. It is not very explicit when that exists early or when it falls through the next step, which is the switch that handles the atlas PE status, which should happen when an atlas PE existed or gets created.

Can we point to the state machine spec somewhere?